cluster/src/main/java/io/atomix/cluster/impl/DefaultClusterMembershipService.java

```diff
@@ -152,118 +157,138 @@ private void handleDiscoveryEvent(NodeDiscoveryEvent event) {
+            }
+
+            PhiAccrualFailureDetector failureDetector = failureDetectors.computeIfAbsent(member.id(), n -> new PhiAccrualFailureDetector());
+            double phi = failureDetector.phi();
+            if (phi >= config.getReachabilityThreshold()
+                || (phi == 0.0 && System.currentTimeMillis() - failureDetector.lastUpdated() > config.getReachabilityTimeout().toMillis())) {
+              if (members.remove(member.id()) != null) {
+                failureDetectors.remove(member.id());
+                post(new ClusterMembershipEvent(ClusterMembershipEvent.Type.MEMBER_REMOVED, member));
+              }
+            }
           }
-        });
+        }, heartbeatScheduler).exceptionally(e -> null)
+        .thenApply(v -> null);
   }
 
   /**
-   * Handles a metadata broadcast from the given address.
-   *
-   * @param address the address from which the metadata was broadcast
-   * @param message the broadcast message
+   * Handles a heartbeat message.
    */
-  private void handleMetadata(Address address, byte[] message) {
+  private byte[] handleHeartbeat(Address address, byte[] message) {
     StatefulMember remoteMember = SERIALIZER.decode(message);
-    LOGGER.trace("{} - Received heartbeat from {}", localMember.id(), remoteMember);
-    StatefulMember localMember = members.get(remoteMember.id());
-    if (localMember != null) {
-      if (!localMember.isReachable()) {
-        LOGGER.info("{} - Member reachable: {}", localMember.id(), localMember);
-        localMember.setReachable(true);
-        post(new ClusterMembershipEvent(ClusterMembershipEvent.Type.REACHABILITY_CHANGED, localMember));
-      }
+    LOGGER.trace("{} - Received heartbeat: {}", localMember.id(), remoteMember);
+    failureDetectors.computeIfAbsent(remoteMember.id(), n -> new PhiAccrualFailureDetector()).report();
+    StatefulMember member = members.get(remoteMember.id());
 
-      if (!Objects.equals(localMember.version(), remoteMember.version()) | !localMember.properties().equals(remoteMember.properties())) {
-        LOGGER.info("{} - Member updated: {}", remoteMember.id(), remoteMember);
+    if (member == null) {
+      remoteMember.setActive(true);
+      remoteMember.setReachable(true);
+      members.put(remoteMember.id(), remoteMember);
+      post(new ClusterMembershipEvent(ClusterMembershipEvent.Type.MEMBER_ADDED, remoteMember));
+    } else {
+      // If the member's metadata changed, overwrite the local member. We do this instead of a version check since
+      // a heartbeat from the source member always takes precedence over a more recent version.
+      if (!Objects.equals(member.version(), remoteMember.version())
+          || !Objects.equals(member.zone(), remoteMember.zone())
+          || !Objects.equals(member.rack(), remoteMember.rack())
+          || !Objects.equals(member.host(), remoteMember.host())
+          || !Objects.equals(member.properties(), remoteMember.properties())) {
```

Comment:
forget which type properties is, should we be using Arrays.equals here or Objects.equals does the job?

Reply:
It's a `Properties` object